### PR TITLE
🔴 Move variable output controls to `_prepare_generation_config `

### PR DESCRIPTION
**Pull request by manueldeprada**
This PR takes the defaults defined in #33870 and now ENFORCES that `prepare_inputs_for_generations()` passes along unused `model_kwargs` to `model_inputs`. This allows to clean up generation loops (and prefill steps). Thus, **it might break Hub-defined old models that do not pass unused arguments**.

Changes:
1. Clean up generation loops by moving `output_attentions` and `output_hidden_states` to `_prepare_generation_config`.
2. Allow these arguments in `_validate_model_kwargs`.
3. Rewrite old models that override `prepare_inputs_for_generation` to follow the standard practice and forward them.
  i. This rewrite makes explicit some weird implicit behaviours: lots of old models are actively ignoring arguments like `attention_mask` or `token_type_ids` and then recreating them in `forward()`

### CODE REVIEW

File: src/transformers/models/xlm_roberta_xl/modeling_xlm_roberta_xl.py [hunk 1]
```diff
@@ -1013,13 +1013,23 @@ def prepare_inputs_for_generation(self, input_ids, past_key_values=None, attenti
...
-        return {
+        model_inputs = {
             "input_ids": input_ids,
             "attention_mask": attention_mask,
             "position_ids": position_ids,
             "past_key_values": past_key_values,
         }
 
+        # They are calculated on the fly on XLMRobertaXLModel.forward()
+        model_kwargs.pop("token_type_ids", None)
```
Comment (manueldeprada) [inline src/transformers/models/xlm_roberta_xl/modeling_xlm_roberta_xl.py:1024]: This is doing nothing new, before it was implicitly done by ignoring model_kwargs, now is explicit. Model fails without this.

**Comment by manueldeprada** [comment]
@gante this was supposed to be small change but the expansive wave got quite big. 

Let me know if I should pursue the PR for cleaner decoding loops (also cleans the prefill PR).

Also, it does not break any public interface, but might break old code on the Hub that have incorrect `prepare_input_for_generations`. Should we mark the PR as 🔴 breaking BC?

**Comment by manueldeprada** [comment]
run-slow: bamba, bloom, ctrl, falcon_h1, falcon_mamba, git, granitemoehybrid, jamba, kosmos2_5, mamba, mamba2, moshi, openai, prophetnet, reformer, rwkv

**Comment by manueldeprada** [comment]
btw checked manually the slow tests failures and they fail on main as well

File: src/transformers/generation/utils.py [hunk 2]
```diff
@@ -1563,7 +1564,7 @@ def _validate_model_kwargs(self, model_kwargs: dict[str, Any]):
                 model_args |= {f"decoder_{x}" for x in decoder_model_args}
 
         for key, value in model_kwargs.items():
-            if value is not None and key not in model_args:
+            if value is not None and key not in model_args and key not in get_type_hints(TransformersKwargs):
```
Comment (gante) [inline src/transformers/generation/utils.py]:
```suggestion
            if value is not None and key not in model_args and key not in TransformersKwargs.__optional_keys__:
```
(then we don't need to import and use `get_type_hints`)

Comment (gante) [reply src/transformers/generation/utils.py]: Also, let's add a comment explaining why we exclude these keys :)

**Review by gante — APPROVED** [review]
LGTM 👍 

Three things:
1. let's make sure to run slow tests on all changed models
2. let's add a test for `prepare_inputs_for_generation` where we test that arbitrary kwargs can be forwarded, but others (like `input_ids`) are manipulated
3. let's take this opportunity to write a better docstring for the base `prepare_inputs_for_generation`. More specifically, let's explain that most non-documented kwargs are simply forwarded.

**Comment by gante** [comment]
@manueldeprada yes, let's add 🔴 (and explain at the top of the PR header why we added it)

**Comment by manueldeprada** [comment]
@gante all done!! Have a look at the test and the new docstring, let me know if you like it!

File: tests/generation/test_utils.py [hunk 3]
```diff
@@ -1835,6 +1835,61 @@ def test_inherits_generation_mixin(self):
...
+                "custom_arg": "test_value",
+                "numeric_arg": 42,
+            }
+
+            model_inputs = model.prepare_inputs_for_generation(**input_args, **arbitrary_kwargs, **extra_kwargs)
+
+            if model_class.__name__ not in non_compilable_model_classes:
+                # Verify that input_ids is cloned
+                input_ids_key = "decoder_input_ids" if config.is_encoder_decoder else "input_ids"
+                self.assertTrue(model_inputs[input_ids_key] is not input_ids)
```
Comment (manueldeprada) [inline tests/generation/test_utils.py]: This is ugly, but fixing all these models not cloning inputs is another rabbithole that I think this PR should avoid right?

Comment (gante) [reply tests/generation/test_utils.py]: Let's try to find other checks instead 👀 Ones that do not need specifying models nor changing modeling code.

E.g.
1. encoder-decoder models should pass the attention mask to `decoder_attention_mask`
2. decoder-only models, when they accept caches (`_supports_default_dynamic_cache()`) and it receives a cache with data, then input_ids should be cropped

This may be a very difficult task -- if these still require custom model logic or overwrites, let's simply test that we can forward arbitrary kwargs
